In [ ]:
#import libraries
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix

In [ ]:
# Read customer churn dataset
df = pd.read_excel('/content/Telco Churn dataset.xlsx')
df.head()

,customerID,gender,SeniorCitizen,MaritalStatus,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,TotalEveMinutes,TotalEveCalls,TotalNightMinutes,TotalNightCalls,TotalIntlMinutes,TotalIntlCalls,CustomerServiceCalls,TotalCall,TotalRevenue,Churn
0,0002-ORFBO,Female,0,Yes,Yes,9,Yes,No,DSL,No,...,241.4,107,204.8,106,15.5,4,0,354,593.3,No
1,0004-TLHLJ,Male,0,No,No,4,Yes,No,Fiber optic,No,...,131.7,94,169.5,106,10.3,9,5,326,280.85,Yes
2,0013-MHZWF,Female,0,No,Yes,9,Yes,No,DSL,No,...,178.6,102,126.8,82,8.0,4,2,324,571.45,No
3,0013-SMEOE,Female,1,Yes,No,71,Yes,No,Fiber optic,Yes,...,262.2,64,130.0,92,8.8,4,0,228,7904.25,No
4,0015-UOCOJ,Female,1,No,No,7,Yes,No,DSL,Yes,...,195.7,86,236.5,137,12.0,2,1,294,340.35,No


# Target Variable Preparation

In [ ]:
# Convert churn labels into numbers
# Yes = 1
# No = 0
df['Churn'] = df['Churn'].map(
    {
        'Yes':1,
        'No': 0
    }
)

# Encode Categorical Variables

In [ ]:
# Convert text categories into numeric columns
df = pd.get_dummies(df, drop_first=True)

# Define Features and Target



In [ ]:
# fEATURES = independent variables
X = df.drop('Churn', axis=1)

# Target = output variables
y = df['Churn']

# Train / Test Split


In [ ]:
# 80% training
# 20% testing
X_train, X_test, y_train, y_test = train_test_split(X,y, test_size=0.2, random_state=42)

# Create decision tree classifier

In [ ]:
tree = DecisionTreeClassifier(max_depth=3, class_weight='balanced', random_state=42)
# max_depth controls tree complexity
# prevents extreme overfitting

# Train

In [ ]:
# model learns splitting rules
tree.fit(X_train, y_train)

DecisionTreeClassifier(class_weight='balanced', max_depth=3, random_state=42)

# Make Predictions

In [ ]:
# predict churn labels on test data
y_pred = tree.predict(X_test)

# Evaluate Model

In [ ]:
# Accuracy
accuracy = accuracy_score(y_test, y_pred)
#Precision
precision = precision_score(y_test, y_pred)
#Recall
recall = recall_score(y_test, y_pred)
# Confusion matrix
cm = confusion_matrix(y_test, y_pred)



In [ ]:
# Print results
print(f'Accuracy: {accuracy}')
print(f'Precision: {precision}')
print(f'Recall: {recall}')
print('Confusion Matrix:')
print(cm)

Accuracy: 0.7421289355322339
Precision: 0.31390134529147984
Recall: 0.7865168539325843
Confusion Matrix:
[[425 153]
 [ 19  70]]


# Model Conclusion

The Decision Tree model achieved:

Accuracy: 74.2%
Precision: 31.4%
Recall: 78.7%

Confusion Matrix:

[[425 153]

 [ 19  70]]

Meaning:

425 customers were correctly predicted as not churning
70 customers were correctly predicted as churning
19 churners were missed
153 non-churners were incorrectly flagged as churners
Business Interpretation

This model is now strongly focused on detecting churners.

The recall of 78.7% means:

The model successfully identifies most customers who are likely to leave.

This is a major improvement compared to the earlier recall of 42%.

For churn prediction, this is often desirable because:

missing churners can lead to revenue loss
businesses usually prefer catching more risky customers
Important Trade-Off

However, precision is relatively low at 31.4%.

This means:

Many customers predicted as churners are actually not going to churn.

This is visible in the confusion matrix:

153 false positives

So the model is aggressive in predicting churn.

Overall Assessment

The model prioritizes:

high recall
over:
high precision

For churn prediction, this is usually acceptable because:

retaining a few extra safe customers is often cheaper
than:
losing actual churners unnoticed
Key Machine Learning Insight

This model demonstrates an important principle:

Improving recall often reduces precision.

After using:

class_weight='balanced'

the model became much better at identifying churners, but it also produced more false alarms.

Final Conclusion

This Decision Tree model is reasonably effective for churn prediction because it captures most at-risk customers.

Although precision is low, the high recall makes the model useful in business scenarios where identifying potential churners is more important than avoiding extra retention outreach.